### PR TITLE
Simple support for tripolar grid topology

### LINKED ISSUE #86: compute tripolar boundary metadata
TODO:

- [ ] add new logic to compute tripolar boundaries
- [ ] add periodicity type enum/flag/string to the metadata netCDF so that we can implement correct boundaries in `nextsim` (match convention with `nextsim` and call it `structure_name`. Check for match in `nextsim`)
- [x] (optional) add decomp args (i.e., command run to generate metadata) to the netCDF. Could be helpful for recomputing metadata (addressed by #88)

### PR DESCRIPTION
**Pull request by Mikolaj-A-Kowalski**
This PR is stacked on #95 and #87

Resolves #86 

Adds a 'least effort' support for the tripolar topology, here I have aimed to minimise any modification to the existing code. 

I am also sneaking a `set -u` in the integration test script since it improves error messages when one tries to run it locally without all the env variables set ;-) 

TODO:
- [ ] Assert preconditions. For example that the `x` size of the tripolar grid must be even. This (and potentially other) assumptions we are relying on need to be verified. I am thinking during the `Grid` construction
- [ ] Remove the `intersection` function. At the moment it is not used (and hence should not be commited yet). I haven't done it yet since getting it out of the history requires more work and I wanted to have this PR opened ASAP 
- [ ] Add more test case. In particular we need one with single rank and 2 domains but with the split that is 'off the centre'.

@TomMelt I think it would be a good to have a quick review of this PR, in particular 2nd pair of eyes to spot some edge cases that might appear would be appreciated. Especially if we can translate them into test. 

I want to hold off polishing this PR to after I can create a asister draft PR in the nextsim which will implement the halo exchange with the tripolar data.

### CODE REVIEW

**Comment by Mikolaj-A-Kowalski** [comment]
The tests passed locally I promise (failing on CI like that is nothing short of a public betrayal! I have a feeling we will be on bad terms going forward...) 

Will try to have a look for the cause of the discrepancy later.

EDIT: 
The reason for failure comes from #87 and #94. The tests should be OK when #87 is updated and we can rebase on the new version.

File: test/integration-test.sh [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-set -e
+set -eu
```
Comment (TomMelt) [inline test/integration-test.sh:3]: thanks for "sneaking" this in :)

File: test/integration-test.sh [hunk 2]
```diff
@@ -55,3 +57,15 @@ for TEST in test_3 test_4 test_4_px test_4_py test_4_px_py; do
...
+  echo "Running integration tripolar test '${TEST}'"
+  ${MPIEXEC} --oversubscribe ${MPIEXEC_NUMPROC_FLAG} 2 ${MPIEXEC_PREFLAGS} \
+    ../decomp -g ${FNAMES[${TEST}]} ${FLAGS[${TEST}]} >/dev/null
+
+  for filename in partition_mask_2 partition_metadata_2; do
+    ncdump "${filename}.nc" >"${filename}.cdl"
+    diff "${filename}.cdl" "${CMAKE_CURRENT_SOURCE_DIR}/${TEST}/ref_${filename}.cdl"
+  done
+  echo -e "\033[0;32mTest passed\033[0m"
+done
```
Comment (TomMelt) [inline test/integration-test.sh:71]: This is absolutely fine and copies the convention we had used above for the previous tests. In future we should perhaps wrap this into a function so we can simplify the script.

Doesn't have to be part of this PR. I kept meaning to do it and now that there 3 of these sections perhaps it's time 😅

File: unit_test/main.cpp
Comment (TomMelt) [inline unit_test/main.cpp:1]: is your plan to remove this file once you have finished writing the other unit tests?

File: unit_test/test_DomainUtils.cpp
Comment (TomMelt) [inline unit_test/test_DomainUtils.cpp:1]: Thanks for adding these much needed unit tests 🙏 it's a great start

File: Grid.cpp [hunk 3]
```diff
@@ -131,14 +132,24 @@ void Grid::ReadGridMask(const std::string& filename, const std::string& mask_nam
...
 
+    // Tripolar grid cannot be periodic
+    if (_tripolar && (_px || _py)) {
+        // TODO:
+        //  - Improve error message
+        //  - Verify conventions our for error handling
+        throw std::invalid_argument(
+            "Tripolar topology and periodic BCs are toggled at the same time");
+    }
+
```
Comment (TomMelt) [inline Grid.cpp:152]: I agree it would be good to set/verify some conventions for error handling. I dont think we followed a specific scheme. So feel free to select something sensible and standardize.

For the message I would be in favor of something like:

"Error: Tripolar BCs are not compatible with other BCs i.e., `--tripolar` is not compatible with `--px` or `--py`."

File: main.cpp [hunk 4]
```diff
@@ -54,7 +54,8 @@ int main(int argc, char* argv[])
         ("ignore-mask,i", po::bool_switch()->default_value(false), "Ignore mask in netCDF grid file")
         ("periodic-x,px", po::bool_switch()->default_value(false), "Periodicity in x-direction")
         ("periodic-y,py", po::bool_switch()->default_value(false), "Periodicity in y-direction")
-        ("output-prefix,op", po::value<string>()->default_value(""), "Prefix for output filenames");
+        ("output-prefix,op", po::value<string>()->default_value(""), "Prefix for output filenames")
+        ("tripolar", po::bool_switch()->default_value(false), "Use tripolar grid topology");
```
Comment (TomMelt) [inline main.cpp:58]: maybe we should mention in the help text that tripolar is not compatible with --px or --py. Or we should change the flag to something like `--boundary-type` and have an enum for `px`, `py`, `px && py`, and `tripolar`.

Comment (TomMelt) [reply main.cpp:58]: This might make the API better by not allowing users to specify incompatible BCs.

File: DomainUtils.hpp [hunk 5]
```diff
@@ -23,7 +23,18 @@ static constexpr std::array<Corner, N_CORNER> corners
...
+
+    // Define arithmetic on Points
+    Point operator+(const Point& other) const { return { x + other.x, y + other.y }; }
+    Point operator-(const Point& other) const { return { x - other.x, y - other.y }; }
+    Point operator*(const int scalar) const { return { x * scalar, y * scalar }; }
+    Point operator-() const { return { -x, -y }; }
+
+    // Implement equality
+    bool operator==(const Point& other) const { return x == other.x && y == other.y; }
+    bool operator!=(const Point& other) const { return !(*this == other); }
```
Comment (TomMelt) [inline DomainUtils.hpp:35]: Thanks for adding the operator overloads. This helps to simplify things :)

File: DomainUtils.hpp [hunk 6]
```diff
@@ -61,4 +115,41 @@ struct Domain {
...
+ *
+ * For a degenerate domain, returns a degenerate domain.
+ * Makes no guarantees about the points inside.
+ * Reflection of an empty set is an empty set is our logic here.
+ *
+ * @param p Point of the symmetry
+ * @param d Domain to be reflected
+ * @return Domain reflected about point p
+ */
+Domain pointReflection(Point p, Domain d);
```
Comment (TomMelt) [inline DomainUtils.hpp:129]: perhaps just me, but it's not clear to me without looking at the implementation what type of reflection this is. Perhaps it is helpful to say it is equivalent to 180 degree rotation about the center point `p`.

File: DomainUtils.hpp [hunk 7]
```diff
@@ -61,4 +115,41 @@ struct Domain {
...
+ * or "point" domain.
+ *
+ * The intersection of disjoint domains returns a degenerate domain that
+ * represents an empty set.
+ *
+ * The intention is to use this function to compute neighbourhood relation
+ * between the domains.
+ *
+ * The operation should be commutative and associative, but we don't test
+ * this property at the moment (albeit we have a bug if it not holds).
```
Comment (TomMelt) [inline DomainUtils.hpp:147]: just a small typo here "we have a bug if it does not hold"

File: DomainUtils.cpp [hunk 8]
```diff
@@ -25,3 +26,64 @@ int domainOverlap(const Domain d1, const Domain d2, const Edge edge)
...
+{
+    Domain d;
+    // Lower left corner
+    d.p1.x = std::min(p1.x, p2.x);
+    d.p1.y = std::min(p1.y, p2.y);
+
+    // Upper right corner
+    d.p2.x = std::max(p1.x, p2.x);
+    d.p2.y = std::max(p1.y, p2.y);
+    return d;
```
Comment (TomMelt) [inline DomainUtils.cpp:40]: Thanks this is a nice way of ensuring that the domain is described correctly i.e., the lower left point is p1.
